File: src/core/contract-ai/evolution/shell-renderer.ts
/**
 * Shell Renderer
 * 
 * @description Renders markdown codeblocks with ANSI colors in terminal.
 *              Provides syntax highlighting for YAML, JSON, TypeScript, JavaScript, Bash, Markdown.
 * 
 * @usage
 *   const renderer = new ShellRenderer();
 *   renderer.codeblock('yaml', 'key: value');
 *   renderer.heading(2, 'Title');
 *   renderer.task('Task name', 'running');
 * 
 * @supported_formats
 *   - yaml, yml     : Key highlighting (cyan), values (green), numbers (magenta)
 *   - json          : Key highlighting (cyan), string values (green), numbers (magenta)
 *   - bash, sh      : Commands (green), comments (gray)
 *   - typescript, ts: Keywords (magenta), strings (green), comments (gray)
 *   - javascript, js: Keywords (magenta), strings (green), comments (gray)
 *   - markdown, md  : Headers (cyan), bold, links
 *   - text, txt, log: Plain white text
 * 
 * @color_scheme
 *   - cyan:    Keys, headers, identifiers
 *   - green:   String values, commands
 *   - magenta: Numbers, keywords
 *   - yellow:  Booleans, warnings
 *   - gray:    Comments, borders, dim text
 *   - red:     Errors
 *   - blue:    Links
 * 
 * @version 1.0.0
 */

// ANSI Color Codes - Standard terminal escape sequences
const COLORS = {
  reset: '\x1b[0m',
  bold: '\x1b[1m',
  dim: '\x1b[2m',
  
  // Foreground
  black: '\x1b[30m',
  red: '\x1b[31m',
  green: '\x1b[32m',
  yellow: '\x1b[33m',
  blue: '\x1b[34m',
  magenta: '\x1b[35m',
  cyan: '\x1b[36m',
  white: '\x1b[37m',
  gray: '\x1b[90m',
  
  // Background
  bgBlack: '\x1b[40m',
  bgBlue: '\x1b[44m',
  bgGray: '\x1b[100m',
};

export class ShellRenderer {
  private useColors: boolean;

  constructor(useColors = true) {
    this.useColors = useColors && process.stdout.isTTY !== false;
  }

  // Color helpers
  private c(color: keyof typeof COLORS, text: string): string {
    if (!this.useColors) return text;
    return `${COLORS[color]}${text}${COLORS.reset}`;
  }

  // Render a heading
  heading(level: number, text: string): void {
    const prefix = '#'.repeat(level);
    console.log(`\n${this.c('bold', this.c('cyan', `${prefix} ${text}`))}\n`);
  }

  // Render a codeblock with syntax highlighting
  codeblock(lang: string, content: string): void {
    const border = this.c('gray', '```' + lang);
    const borderEnd = this.c('gray', '```');
    
    console.log(border);
    
    const lines = content.split('\n');
    for (const line of lines) {
      console.log(this.highlightLine(lang, line));
    }
    
    console.log(borderEnd);
  }

  // Highlight a single line based on language
  private highlightLine(lang: string, line: string): string {
    switch (lang.toLowerCase()) {
      case 'yaml':
      case 'yml':
        return this.highlightYaml(line);
      case 'json':
        return this.highlightJson(line);
      case 'bash':
      case 'sh':
      case 'shell':
        return this.highlightBash(line);
      case 'typescript':
      case 'ts':
      case 'javascript':
      case 'js':
        return this.highlightJs(line);
      case 'text':
      case 'txt':
      case 'log':
        return this.c('white', line);
      case 'markdown':
      case 'md':
        return this.highlightMarkdown(line);
      default:
        return this.c('white', line);
    }
  }

  // JavaScript/TypeScript syntax highlighting
  private highlightJs(line: string): string {
    let result = line;
    
    // Comments
    if (line.trim().startsWith('//')) {
      return this.c('gray', line);
    }
    
    // Keywords
    const keywords = ['const', 'let', 'var', 'function', 'async', 'await', 'return', 'if', 'else', 'for', 'while', 'import', 'export', 'from', 'class', 'interface', 'type'];
    for (const kw of keywords) {
      result = result.replace(new RegExp(`\\b${kw}\\b`, 'g'), this.c('magenta', kw));
    }
    
    // Strings
    result = result.replace(/'([^']*)'/g, this.c('green', "'$1'"));
    result = result.replace(/"([^"]*)"/g, this.c('green', '"$1"'));
    result = result.replace(/`([^`]*)`/g, this.c('green', '`$1`'));
    
    return result;
  }

  // Markdown syntax highlighting
  private highlightMarkdown(line: string): string {
    // Headers
    if (line.match(/^#{1,6}\s/)) {
      return this.c('cyan', line);
    }
    // Bold
    if (line.includes('**')) {
      return line.replace(/\*\*([^*]+)\*\*/g, this.c('bold', '$1'));
    }
    // Links
    if (line.includes('[')) {
      return line.replace(/\[([^\]]+)\]\(([^)]+)\)/g, `${this.c('blue', '[$1]')}(${this.c('gray', '$2')})`);
    }
    return line;
  }

  // YAML syntax highlighting
  private highlightYaml(line: string): string {
    // Comments
    if (line.trim().startsWith('#')) {
      return this.c('gray', line);
    }
    
    // Key: value pairs
    const keyMatch = line.match(/^(\s*)([^:]+)(:)(.*)$/);
    if (keyMatch) {
      const [, indent, key, colon, value] = keyMatch;
      const coloredKey = this.c('cyan', key);
      const coloredColon = this.c('white', colon);
      let coloredValue = value;
      
      // Color values based on type
      const trimmedValue = value.trim();
      if (trimmedValue === 'true' || trimmedValue === 'false') {
        coloredValue = this.c('yellow', value);
      } else if (trimmedValue.match(/^\d+$/)) {
        coloredValue = this.c('magenta', value);
      } else if (trimmedValue.startsWith('"') || trimmedValue.startsWith("'")) {
        coloredValue = this.c('green', value);
      } else if (trimmedValue === 'null' || trimmedValue === '~') {
        coloredValue = this.c('gray', value);
      } else {
        coloredValue = this.c('green', value);
      }
      
      return `${indent}${coloredKey}${coloredColon}${coloredValue}`;
    }
    
    // List items
    if (line.trim().startsWith('-')) {
      const match = line.match(/^(\s*)(-)(.*)$/);
      if (match) {
        const [, indent, dash, rest] = match;
        return `${indent}${this.c('white', dash)}${this.c('green', rest)}`;
      }
    }
    
    return line;
  }

  // JSON syntax highlighting
  private highlightJson(line: string): string {
    // Simple JSON highlighting
    let result = line;
    
    // Keys
    result = result.replace(/"([^"]+)":/g, (_, key) => 
      `${this.c('cyan', `"${key}"`)}:`
    );
    
    // String values
    result = result.replace(/: "([^"]+)"/g, (_, val) => 
      `: ${this.c('green', `"${val}"`)}`
    );
    
    // Numbers
    result = result.replace(/: (\d+)/g, (_, num) => 
      `: ${this.c('magenta', num)}`
    );
    
    // Booleans
    result = result.replace(/: (true|false)/g, (_, bool) => 
      `: ${this.c('yellow', bool)}`
    );
    
    // null
    result = result.replace(/: null/g, `: ${this.c('gray', 'null')}`);
    
    return result;
  }

  // Bash syntax highlighting
  private highlightBash(line: string): string {
    // Comments
    if (line.trim().startsWith('#')) {
      return this.c('gray', line);
    }
    
    // Commands
    const words = line.split(' ');
    if (words.length > 0) {
      const cmd = words[0];
      const args = words.slice(1).join(' ');
      return `${this.c('green', cmd)} ${this.c('white', args)}`;
    }
    
    return line;
  }

  // Render status icons with colors
  status(icon: string, message: string, type: 'info' | 'success' | 'warning' | 'error' = 'info'): void {
    const colors: Record<string, keyof typeof COLORS> = {
      info: 'blue',
      success: 'green',
      warning: 'yellow',
      error: 'red'
    };
    console.log(`${icon} ${this.c(colors[type], message)}`);
  }

  // Render a key-value line
  kv(key: string, value: string | number | boolean): void {
    console.log(`  ${this.c('cyan', key)}: ${this.c('white', String(value))}`);
  }

  // Render progress
  progress(done: number, total: number, label?: string): void {
    const pct = Math.round((done / total) * 100);
    const bar = '█'.repeat(Math.floor(pct / 5)) + '░'.repeat(20 - Math.floor(pct / 5));
    const labelStr = label ? ` ${label}` : '';
    console.log(`${this.c('gray', '[')}${this.c('green', bar)}${this.c('gray', ']')} ${this.c('cyan', `${pct}%`)}${labelStr}`);
  }

  // Render a separator
  separator(): void {
    console.log(this.c('gray', '─'.repeat(60)));
  }

  // Render inline code
  inline(text: string): string {
    return this.c('cyan', text);
  }

  // Render a task status
  task(name: string, status: 'pending' | 'running' | 'done' | 'failed', duration?: number): void {
    const icons: Record<string, string> = {
      pending: '⏳',
      running: '🔄',
      done: '✅',
      failed: '❌'
    };
    const colors: Record<string, keyof typeof COLORS> = {
      pending: 'gray',
      running: 'yellow',
      done: 'green',
      failed: 'red'
    };
    
    const durationStr = duration !== undefined ? this.c('gray', ` (${duration}s)`) : '';
    console.log(`${icons[status]} ${this.c(colors[status], name)}${durationStr}`);
  }
}

// Singleton instance
let renderer: ShellRenderer | null = null;

export function getRenderer(): ShellRenderer {
  if (!renderer) {
    renderer = new ShellRenderer();
  }
  return renderer;
}

// Convenience functions
export const render = {
  heading: (level: number, text: string) => getRenderer().heading(level, text),
  code: (lang: string, content: string) => getRenderer().codeblock(lang, content),
  status: (icon: string, msg: string, type?: 'info' | 'success' | 'warning' | 'error') => getRenderer().status(icon, msg, type),
  kv: (key: string, value: string | number | boolean) => getRenderer().kv(key, value),
  progress: (done: number, total: number, label?: string) => getRenderer().progress(done, total, label),
  separator: () => getRenderer().separator(),
  task: (name: string, status: 'pending' | 'running' | 'done' | 'failed', duration?: number) => getRenderer().task(name, status, duration),
  inline: (text: string) => getRenderer().inline(text),
};

// Helper to wrap console.log with markdown rendering
export function renderMarkdown(text: string): void {
  const r = getRenderer();
  const lines = text.split('\n');
  let inCodeblock = false;
  let codeblockLang = '';
  let codeblockContent: string[] = [];

  for (const line of lines) {
    // Check for codeblock start/end
    if (line.startsWith('```')) {
      if (inCodeblock) {
        // End of codeblock
        r.codeblock(codeblockLang, codeblockContent.join('\n'));
        inCodeblock = false;
        codeblockLang = '';
        codeblockContent = [];
      } else {
        // Start of codeblock
        inCodeblock = true;
        codeblockLang = line.slice(3).trim();
      }
      continue;
    }

    if (inCodeblock) {
      codeblockContent.push(line);
      continue;
    }

    // Headings
    const headingMatch = line.match(/^(#{1,6})\s+(.+)$/);
    if (headingMatch) {
      r.heading(headingMatch[1].length, headingMatch[2]);
      continue;
    }

    // Regular text
    console.log(line);
  }
}

File: src/python/reclapp/evolution/shell_renderer.py
"""
Shell Renderer

Renders colorized markdown output in terminal.
Uses clickmd.MarkdownRenderer as rendering backend.

Mirrors: src/core/contract-ai/evolution/shell-renderer.ts
@version 2.4.1 - Unified with clickmd
"""

import re
import sys
from typing import Literal, Optional

# Import clickmd renderer (or fallback to local ANSI codes)
try:
    from clickmd.renderer import MarkdownRenderer as _ClickmdRenderer, _COLORS as COLORS
    CLICKMD_AVAILABLE = True
except ImportError:
    CLICKMD_AVAILABLE = False
    _ClickmdRenderer = None
    # Fallback ANSI color codes
    COLORS = {
        "reset": "\033[0m",
        "bold": "\033[1m",
        "dim": "\033[2m",
        "cyan": "\033[36m",
        "green": "\033[32m",
        "yellow": "\033[33m",
        "magenta": "\033[35m",
        "red": "\033[31m",
        "blue": "\033[34m",
        "gray": "\033[90m",
        "white": "\033[37m",
    }


Language = Literal["yaml", "json", "bash", "typescript", "javascript", "markdown", "log", "text"]


class ShellRenderer:
    """
    Renders colorized output in terminal.
    
    Uses clickmd.MarkdownRenderer when available for consistent rendering
    across all reclapp components.
    
    Supports syntax highlighting for:
    - YAML
    - JSON
    - Bash
    - TypeScript/JavaScript
    - Markdown
    - Log messages
    
    Example:
        renderer = ShellRenderer(verbose=True)
        renderer.heading(2, "Status")
        renderer.codeblock("yaml", "status: ok\\ncount: 42")
    """
    
    def __init__(self, verbose: bool = True):
        self.verbose = verbose
        self._clickmd_renderer: Optional[_ClickmdRenderer] = None
        self._log_enabled = False
        self._log_buffer: list[str] = []
        
        # Initialize clickmd renderer if available
        if CLICKMD_AVAILABLE and _ClickmdRenderer is not None:
            self._clickmd_renderer = _ClickmdRenderer(use_colors=True, stream=sys.stdout)

    @property
    def renderer(self) -> Optional["_ClickmdRenderer"]:
        """Expose underlying clickmd renderer"""
        return self._clickmd_renderer

    def enable_log(self) -> None:
        """Enable log buffering for markdown export"""
        self._log_enabled = True
        self._log_buffer = []

    def get_log(self) -> str:
        """Get buffered log as clean markdown"""
        return "\n".join(self._log_buffer)

    def clear_log(self) -> None:
        """Clear log buffer"""
        self._log_buffer = []

    def save_log(self, file_path: str) -> None:
        """Save log to file as markdown"""
        import os
        dir_path = os.path.dirname(file_path)
        if dir_path and not os.path.exists(dir_path):
            os.makedirs(dir_path, exist_ok=True)
        with open(file_path, "w", encoding="utf-8") as f:
            f.write(self.get_log())

    def _log(self, text: str) -> None:
        """Buffer log line if logging is enabled"""
        if self._log_enabled:
            # Strip ANSI codes for clean markdown
            clean = re.sub(r"\x1b\[[0-9;]*m", "", text)
            self._log_buffer.append(clean)

    def _get_clickmd(self):
        """Get clickmd module (lazy load)"""
        if hasattr(self, '_clickmd_module'):
            return self._clickmd_module
        try:
            import clickmd
            self._clickmd_module = clickmd
        except Exception:
            self._clickmd_module = None
        return self._clickmd_module

    def _try_md(self, text: str) -> bool:
        clickmd = self._get_clickmd()
        self._log(text)  # Buffer for log file
        if not clickmd:
            return False
        clickmd.md(text)
        return True

    def _try_echo(self, text: str) -> bool:
        clickmd = self._get_clickmd()
        self._log(text)  # Buffer for log file
        if not clickmd:
            return False
        clickmd.echo(text)
        return True
    
    def heading(self, level: int, text: str) -> None:
        """Print a heading"""
        if not self.verbose:
            return

        if self._try_md(f"{'#' * level} {text}\n"):
            return
        
        prefix = "#" * level
        self._log(f"{prefix} {text}")
        print(f"\n{COLORS['bold']}{COLORS['cyan']}{prefix} {text}{COLORS['reset']}\n")
    
    def codeblock(self, language: Language, content: str) -> None:
        """Print a syntax-highlighted code block"""
        if not self.verbose:
            return

        if self._try_md(f"```{language}\n{content}\n```\n"):
            return
        
        self._log(f"```{language}")
        print(f"{COLORS['dim']}```{language}{COLORS['reset']}")
        
        for line in content.split("\n"):
            self._log(line)
            highlighted = self._highlight_line(line, language)
            print(highlighted)
        
        self._log("```")
        print(f"{COLORS['dim']}```{COLORS['reset']}")
    
    def text(self, content: str) -> None:
        """Print plain text"""
        if not self.verbose:
            return

        if self._try_echo(content):
            return
        self._log(content)
        print(content)
    
    def success(self, message: str) -> None:
        """Print success message"""
        if not self.verbose:
            return

        if self._try_md(f"```log\n✅ {message}\n```\n"):
            return
        self._log(f"✅ {message}")
        print(f"{COLORS['green']}✅ {message}{COLORS['reset']}")
    
    def error(self, message: str) -> None:
        """Print error message"""
        if not self.verbose:
            return

        if self._try_md(f"```log\n❌ {message}\n```\n"):
            return
        self._log(f"❌ {message}")
        print(f"{COLORS['red']}❌ {message}{COLORS['reset']}")
    
    def warning(self, message: str) -> None:
        """Print warning message"""
        if not self.verbose:
            return

        if self._try_md(f"```log\n⚠️ {message}\n```\n"):
            return
        self._log(f"⚠️ {message}")
        print(f"{COLORS['yellow']}⚠️ {message}{COLORS['reset']}")
    
    def info(self, message: str) -> None:
        """Print info message"""
        if not self.verbose:
            return

        if self._try_md(f"```log\n→ {message}\n```\n"):
            return
        self._log(f"→ {message}")
        print(f"{COLORS['cyan']}ℹ️ {message}{COLORS['reset']}")
    
    def _highlight_line(self, line: str, language: Language) -> str:
        """Apply syntax highlighting to a line"""
        if language == "yaml":
            return self._highlight_yaml(line)
        elif language == "json":
            return self._highlight_json(line)
        elif language == "bash":
            return self._highlight_bash(line)
        elif language in ("typescript", "javascript"):
            return self._highlight_js(line)
        elif language == "log":
            return self._highlight_log(line)
        else:
            return line
    
    def _highlight_yaml(self, line: str) -> str:
        """Highlight YAML syntax"""
        # Comments
        if line.strip().startswith("#"):
            return f"{COLORS['gray']}{line}{COLORS['reset']}"
        
        # Key: value
        match = re.match(r'^(\s*)([^:]+)(:)(.*)$', line)
        if match:
            indent, key, colon, value = match.groups()
            
            # Colorize value based on type
            value_colored = value
            if value.strip().isdigit() or re.match(r'^\s*-?\d+\.?\d*$', value):
                value_colored = f"{COLORS['magenta']}{value}{COLORS['reset']}"
            elif value.strip().lower() in ("true", "false", "yes", "no", "null", "none"):
                value_colored = f"{COLORS['yellow']}{value}{COLORS['reset']}"
            elif value.strip().startswith('"') or value.strip().startswith("'"):
                value_colored = f"{COLORS['green']}{value}{COLORS['reset']}"
            elif value.strip():
                value_colored = f"{COLORS['green']}{value}{COLORS['reset']}"
            
            return f"{indent}{COLORS['cyan']}{key}{COLORS['reset']}{colon}{value_colored}"
        
        # List item
        if line.strip().startswith("-"):
            return f"{COLORS['yellow']}{line}{COLORS['reset']}"
        
        return line
    
    def _highlight_json(self, line: str) -> str:
        """Highlight JSON syntax"""
        # Keys
        line = re.sub(
            r'"([^"]+)"\s*:',
            f'{COLORS["cyan"]}"\\1"{COLORS["reset"]}:',
            line
        )
        # String values
        line = re.sub(
            r':\s*"([^"]*)"',
            f': {COLORS["green"]}"\\1"{COLORS["reset"]}',
            line
        )
        # Numbers
        line = re.sub(
            r':\s*(\d+\.?\d*)',
            f': {COLORS["magenta"]}\\1{COLORS["reset"]}',
            line
        )
        # Booleans
        line = re.sub(
            r':\s*(true|false|null)',
            f': {COLORS["yellow"]}\\1{COLORS["reset"]}',
            line
        )
        return line
    
    def _highlight_bash(self, line: str) -> str:
        """Highlight Bash syntax"""
        # Comments
        if line.strip().startswith("#"):
            return f"{COLORS['gray']}{line}{COLORS['reset']}"
        
        # Commands
        commands = ["cd", "npm", "node", "python", "pip", "git", "docker", "reclapp", "ollama"]
        for cmd in commands:
            if line.strip().startswith(cmd):
                return f"{COLORS['cyan']}{line}{COLORS['reset']}"
        
        return line
    
    def _highlight_js(self, line: str) -> str:
        """Highlight JavaScript/TypeScript syntax"""
        # Keywords
        keywords = ["const", "let", "var", "function", "async", "await", "return", "import", "export", "from", "class", "interface", "type"]
        for kw in keywords:
            line = re.sub(
                rf'\b{kw}\b',
                f'{COLORS["magenta"]}{kw}{COLORS["reset"]}',
                line
            )
        
        # Strings
        line = re.sub(
            r'(["\'])(.*?)\1',
            f'{COLORS["green"]}\\1\\2\\1{COLORS["reset"]}',
            line
        )
        
        # Comments
        if "//" in line:
            parts = line.split("//", 1)
            line = parts[0] + f"{COLORS['gray']}//{parts[1]}{COLORS['reset']}"
        
        return line
    
    def _highlight_log(self, line: str) -> str:
        """Highlight log messages"""
        if "✅" in line or "success" in line.lower():
            return f"{COLORS['green']}{line}{COLORS['reset']}"
        elif "❌" in line or "error" in line.lower() or "fail" in line.lower():
            return f"{COLORS['red']}{line}{COLORS['reset']}"
        elif "⚠️" in line or "warn" in line.lower():
            return f"{COLORS['yellow']}{line}{COLORS['reset']}"
        elif "ℹ️" in line or "info" in line.lower():
            return f"{COLORS['cyan']}{line}{COLORS['reset']}"
        return line
